### ISSUE TITLE
the performance of tensorflow distributed 

### ISSUE THREAD

**Issue by dingevin** [tensorflow/tensorflow#24034]
I have trained a speech recognition network using tensorflow both on multi-gpu and distributed. there is some question that i have been puzzled a long time. that's my [code,](https://github.com/dingevin/distributed-training) when i'm training i faced two questions:
1. **the accuracy of distributed not equal to multi-gpu**. in multi-gpu version, i use 2 gpu and training 8000 steps, the accuracy can reach 32%; however the accuracy only can reach 28% at distributed version(i have 2 ps 2 worker, and each worker use 2 gpu), each worker training 4000 steps, and the learning rate was same. it seems the parameters not share.
multi-gpu: (when training 4000 steps， loss is reduce to about 5.0)
```
INFO:tensorflow:Step #4015: lr: 0.000100, time 2018-11-28 16:35:16, accuracy 17.12786%, cross entropy 5.230135(147.0 examples/sec; 0.218 sec/batch)
INFO:tensorflow:Step #4016: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 16.93807%, cross entropy 5.064457(151.4 examples/sec; 0.211 sec/batch)
INFO:tensorflow:Step #4017: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 18.51658%, cross entropy 5.146353(149.4 examples/sec; 0.214 sec/batch)
INFO:tensorflow:Step #4018: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 18.98760%, cross entropy 4.905624(145.8 examples/sec; 0.219 sec/batch)
INFO:tensorflow:Step #4019: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 19.00991%, cross entropy 4.929277(139.1 examples/sec; 0.230 sec/batch)
INFO:tensorflow:Step #4020: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 15.84382%, cross entropy 5.140810(139.6 examples/sec; 0.229 sec/batch)
INFO:tensorflow:Step #4021: lr: 0.000100, time 2018-11-28 16:35:18, accuracy 20.37208%, cross entropy 4.693164(143.1 examples/sec; 0.224 sec/batch)
```
distributed ( training 2000 steps， the loss was reduce to 5.5， equal to 2000 steps at multi-gpu )
```
INFO:tensorflow:Step #2018: learning_rate:0.0001, accuracy 17.56266%, cross entropy 5.382931(15.6 examples/sec; 2.050 sec/batch)
INFO:tensorflow:Step #2019: learning_rate:0.0001, accuracy 16.61609%, cross entropy 5.386977(21.3 examples/sec; 1.505 sec/batch)
INFO:tensorflow:Step #2020: learning_rate:0.0001, accuracy 11.97795%, cross entropy 6.026215(29.3 examples/sec; 1.093 sec/batch)
INFO:tensorflow:Step #2021: learning_rate:0.0001, accuracy 14.76162%, cross entropy 5.657806(32.7 examples/sec; 0.977 sec/batch)
INFO:tensorflow:Step #2022: learning_rate:0.0001, accuracy 15.25826%, cross entropy 5.588447(29.4 examples/sec; 1.087 sec/batch)
INFO:tensorflow:Step #2023: learning_rate:0.0001, accuracy 13.63419%, cross entropy 5.882462(28.7 examples/sec; 1.117 sec/batch)
INFO:tensorflow:Step #2024: learning_rate:0.0001, accuracy 15.66866%, cross entropy 5.585463(37.7 examples/sec; 0.849 sec/batch)
```
2. **the speed of distributed version was so slow**. the bandwidth is 125Mb/s, is not the bottleneck.
i execute the command as follow on every worker:
```
CUDA_VISIBLE_DEVICES='' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220, gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=ps --task_index=0 &
CUDA_VISIBLE_DEVICES='0,1' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220,gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=worker --task_index=0 --gpu_num=2 --learning_rate=0.0001--how_many_training_steps=4000 --save_step_interval=5000 &
```
and another worker:
```
CUDA_VISIBLE_DEVICES='' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220, gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=ps --task_index=1 &
CUDA_VISIBLE_DEVICES='0,1' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220,gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=worker --task_index=1 --gpu_num=2 --learning_rate=0.0001--how_many_training_steps=4000 --save_step_interval=5000 &
```

was my code wrong? or i misunderstand distributed? can somebody give me some advice. thanks.

**Comment by byronyi** [comment]
__the accuracy of distributed not equal to multi-gpu__: usually the case. Considering your batch size doubles if you use 2x number of GPU. You need to adjust your learning rate accordingly to reach the same convergence level.

__bandwidth is 125Mb/s, is not the bottleneck__: seems too rough. What's is your model size and how long does each step take? You may use networking bandwidth monitoring tools like `jnettop` or `collectl` to see your actual network traffic.

PS: we usually use at least 10Gbps network for distributed DL, which is 10x than that of your environment.

**Comment by byronyi** [comment]
I will suggest you to keep the batch size and number of training steps same in both cases before you trying to track down the issue to distributed TF.

For example, if you use batch size 128 per GPU in 2 GPU case (256 in total) then you could try 64 per GPU in 4 GPU (256 in total) case and keep your lr and number of epoch the same. 

Linear scaling is fine, but it doesn’t mean you get exactly the same results.

**Comment by harshini-gadige** [comment]
Closing this as it is in "awaiting response" status for more than 7 days. Feel free to add your comments and we will reopen(if required).
